source/distributed/tria.cc

```diff
@@ -1124,10 +1124,545 @@ namespace
+                  CELL_INVALID:
+                  // do nothing on invalid cells
+                  break;
+
+                default:
+                  Assert(false, ExcInternalError());
+                  break;
+              }
+#  endif
+
+            // Reserve memory corresponding to the amount of callback
+            // function that will be called.
+            // On cells flagged with CELL_INVALID, only its CellStatus
+            // will be stored.
+            const unsigned int num_callbacks_on_cell =
+              1 +
+              ((cell_status ==
+                parallel::distributed::Triangulation<dim,
+                                                     spacedim>::CELL_INVALID) ?
+                 0 :
+                 pack_callbacks.size());
+            it_data_cell->resize(num_callbacks_on_cell);
+
+            // We continue with packing all data on this specific cell.
+            // First, we pack the CellStatus information.
+            auto it_data_callback = it_data_cell->begin();
+            *it_data_callback =
+              Utilities::pack(cell_status, /*allow_compression=*/false);
+
+            // Proceed with all registered callback functions.
+            // Skip cells with the CELL_INVALID flag.
+            if (cell_status !=
+                parallel::distributed::Triangulation<dim,
+                                                     spacedim>::CELL_INVALID)
+              {
+                ++it_data_callback;
+                for (auto cit_callback = pack_callbacks.cbegin();
+                     cit_callback != pack_callbacks.cend();
+                     ++cit_callback, ++it_data_callback)
+                  {
+                    (*cit_callback)(dealii_cell,
+                                    cell_status,
+                                    *it_data_callback);
+                  }
+              }
+          } // loop over quad_cell_relations
+      }
+
+      // Generate a vector which stores the sizes of each callback function,
+      // including the packed CellStatus transfer.
+      // Find the very first cell that we wrote to with all callback
+      // functions (i.e. a cell that was not flagged with CELL_INVALID)
+      // and store the sizes of each buffer.
+      std::vector<unsigned int> local_sizes_fix(1 + pack_callbacks.size());
+      for (auto cit_data_cell = packed_data.cbegin();
+           cit_data_cell != packed_data.cend();
+           ++cit_data_cell)
+        {
+          if (cit_data_cell->size() == local_sizes_fix.size())
+            {
+              auto it_sizes_fix      = local_sizes_fix.begin();
+              auto cit_data_callback = cit_data_cell->cbegin();
+              for (; cit_data_callback != cit_data_cell->cend();
+                   ++cit_data_callback, ++it_sizes_fix)
+                {
+                  *it_sizes_fix = cit_data_callback->size();
+                }
+
+              break;
```

Comment:
did we convince ourselves that each processor has at least one of these cells? if not, is the following well-defined?

in any case, leave a comment somewhere that pertains to this question